docs: expand error codes documentation with reasons and debugging steps (#1006)
safe-global/safe-smart-account: docs/error_codes.md

## Error codes

### General init related
- `GS000`: Could not finish initialization

>   - **Why:** Delegatecall to a module during initialization failed (see `setupModules`).
>   - **How to debug/solve:** Ensure the module address is correct and the called contract does not revert.

Reviewer: We've called this the "initializer" in the past, as this contract isn't really a module. I think it is done in the `setupModules` as an implementation detail, but I would keep it out of the help.

```suggestion
  - **Why:** Delegatecall to the initializer during setup failed.
  - **How to debug/solve:** Ensure the contract address is correct and does not revert.
```

- `GS001`: Threshold needs to be defined
  - **Why:** The signature threshold is not set (equals 0) when checking signatures.

>   - **How to debug/solve:** Make sure the Safe is properly initialized and the threshold is greater than 0.

Reviewer: ```suggestion
  - **How to debug/solve:** Make sure the Safe is properly initialized.
```

Rationale: initialization (setup) ensures `threshold != 0`.

  - **Why:** Attempt to initialize a module at an address that is not a contract.
  - **How to debug/solve:** Check that the module address points to a deployed contract.

### General gas/execution related
- `GS010`: Not enough gas to execute Safe transaction
  - **Why:** `execTransaction` was called with insufficient gas.
  - **How to debug/solve:** Increase the gas limit for the transaction.
- `GS011`: Could not pay gas costs with ether
  - **Why:** Payment for gas in ETH failed (e.g., recipient cannot receive ETH).
  - **How to debug/solve:** Ensure the receiver address is correct and can accept ETH.
- `GS012`: Could not pay gas costs with token
  - **Why:** Payment for gas in token failed (e.g., token does not support transfer or returns false).
  - **How to debug/solve:** Check that the token supports the transfer standard and the Safe has enough balance.
- `GS013`: Safe transaction failed when gasPrice and safeTxGas were 0 (Deprecated in v1.5.0)
  - **Why:** Migration or similar function requiring delegatecall was called outside of delegatecall context or with invalid parameters.

>   - **How to debug/solve:** Only use supported migration scenarios and ensure the call is made via delegatecall.

Reviewer: This happens on general transaction revert, and is not specific to delegatecall.

  - **How to debug/solve:** Ensure the number of signatures matches the threshold.
- `GS021`: Invalid contract signature location: inside static part
  - **Why:** The signature offset points inside the static part of the data, not the dynamic part.
  - **How to debug/solve:** Check the correctness of the signature array encoding.
- `GS022`: Invalid contract signature location: length not present
  - **Why:** The signature offset is out of bounds (no length for the signature).
  - **How to debug/solve:** Ensure the signature offset does not exceed the array length.
- `GS023`: Invalid contract signature location: data not complete
  - **Why:** The signature data does not fully fit within the signatures array.
  - **How to debug/solve:** Ensure the signature length and offset are correct.
- `GS024`: Invalid contract signature provided
  - **Why:** The contract signer returned an invalid value (not EIP1271_MAGIC_VALUE) during signature validation.
  - **How to debug/solve:** Check that the contract signer implements EIP-1271 and returns the correct value.
- `GS025`: Hash has not been approved
  - **Why:** The transaction hash was not approved by the required number of owners.

>   - **How to debug/solve:** Ensure all required owners have signed the transaction or called approveHash.

Reviewer: ```suggestion
  - **How to debug/solve:** Ensure the owner called the Safe directory or previously approved the has (with `approveHash`).
```

- `GS026`: Invalid owner provided
  - **Why:** An invalid or duplicate owner was found in the signatures array.
  - **How to debug/solve:** Ensure each owner is unique and valid.

### General auth related
- `GS030`: Only owners can approve a hash
  - **Why:** Attempt to approve a hash by a non-owner.

>   - **How to debug/solve:** Only owner addresses should be used for approveHash.

Reviewer: ```suggestion
  - **How to debug/solve:** Only owners can call `approveHash`.
```

  - **How to debug/solve:** Increase the gas limit for the transaction.
- `GS011`: Could not pay gas costs with ether
  - **Why:** Payment for gas in ETH failed (e.g., recipient cannot receive ETH, or the Safe does not have enough funds).
  - **How to debug/solve:** Ensure the receiver address is correct and can accept ETH, and that the Safe has sufficient balance to cover the payment.
- `GS012`: Could not pay gas costs with token
  - **Why:** Payment for gas in token failed (e.g., token does not support transfer or returns false).
  - **How to debug/solve:** Check that the token supports the transfer standard and the Safe has enough balance.
- `GS013`: Safe transaction reverted (Deprecated in v1.5.0)
  - **Why:** The transaction reverted for any reason (not limited to delegatecall or migration scenarios). This is a general revert error for failed transactions when `gasPrice` and `safeTxGas` are 0.
  - **How to debug/solve:** Check the transaction parameters, contract logic, and ensure all preconditions are met. Review the revert reason for more details.

### General signature validation related
- `GS020`: Signatures data too short

>   - **Why:** Not enough data was provided for signature validation (less than threshold * 65 bytes).

Reviewer: ```suggestion
  - **Why:** Not enough data was provided for signature validation (less than `threshold * 65` bytes).
```

  - **How to debug/solve:** Ensure the number of signatures matches the threshold.
- `GS021`: Invalid contract signature location: inside static part
  - **Why:** The signature offset points inside the static part of the data, not the dynamic part.
  - **How to debug/solve:** Check the signatures are correctly encoded.
- `GS022`: Invalid contract signature location: length not present
  - **Why:** The signature offset is out of bounds (no length for a contract signature).
  - **How to debug/solve:** Ensure the signature offset does not exceed the array length.
- `GS023`: Invalid contract signature location: data not complete
  - **Why:** The signature data does not fully fit within the signatures array.
  - **How to debug/solve:** Ensure the signature length and offset are correct.
- `GS024`: Invalid contract signature provided
  - **Why:** The contract signer returned an invalid value (not EIP1271_MAGIC_VALUE) during signature validation.
  - **How to debug/solve:** Check that the contract signer implements EIP-1271 and returns the correct value.
- `GS025`: Hash has not been approved
  - **Why:** The transaction hash was not approved by the required number of owners.

>   - **How to debug/solve:** Ensure the owner called the Safe directory or previously approved the has (with `approveHash`).

Reviewer 2: Also this should say `[...] called the Safe directly [...]`, right? (cc @nlordell )
